### ISSUE TITLE
kubernetes-the-hard-way  with Docker-ce

### ISSUE THREAD

**Issue by kylix3511** [kelseyhightower/kubernetes-the-hard-way#401]
Did any one try — Kelsey Hightower @ kubernetes-the-hard-way with Docker-ce ( instead of containerd ) ? 

I kind of curious and interested to know and start. ( But don't know where to start ) ?

**Comment by kylix3511** [comment]
Thanks @estesp ... for sharing the link about docker-ce in kubernetes-the-hard-way --- https://github.com/kelseyhightower/kubernetes-the-hard-way/blob/ee5790bc155d3645e7599e1f436a5ee5aceb1f48/docs/06-kubernetes-worker.md (edited)

It's really helpful and useful. I experiment it. 

In my lab, I did work to replace the containerd with Docker in the Kelsey Hightower — Kubernetes the hardway setup/guide. 

The docker installation was smooth and follow the similar process …  

But the Worker’s kubelet is not starting ..

``` 

Every 2.0s: systemctl status kubelet                                                     dworker1.mylabserver.com: Mon Oct 29 12:35:40 2018
● kubelet.service - Kubernetes Kubelet
   Loaded: loaded (/etc/systemd/system/kubelet.service; enabled; vendor preset: enabled)
   Active: activating (auto-restart) (Result: exit-code) since Mon 2018-10-29 12:35:36 PDT; 3s ago
     Docs: https://github.com/kubernetes/kubernetes
  Process: 20716 ExecStart=/usr/local/bin/kubelet --config=/var/lib/kubelet/kubelet-config.yaml --container-runtime=remote --allow-privileg
ed=true --kubeconfig=/var/lib/kubelet/kubeconfig --container-runtime-endpoint=unix:///var/run/docker.sock --image-pull-progress-deadline=2m
 --register-node=true --hostname-override=dworker1.mylabserver.com --network-plugin=cni --ignore-preflight-errors cri --cni-conf-dir=/etc/c
ni/net.d --v=2 (code=exited, status=255)
 Main PID: 20716 (code=exited, status=255)
Oct 29 12:35:36 dworker1.mylabserver.com kubelet[20716]:       --runonce                                             If true, exit after sp
awning pods from static pod files or remote urls. Exclusive with --enable-server
Oct 29 12:35:36 dworker1.mylabserver.com kubelet[20716]:       --runtime-cgroups string                              Optional absolute name
 of cgroups to create and run the runtime in.
Oct 29 12:35:36 dworker1.mylabserver.com kubelet[20716]:       --seccomp-profile-root string                         <Warning: Alpha featur
e> Directory path for seccomp profiles. (default "/var/lib/kubelet/seccomp")
Oct 29 12:35:36 dworker1.mylabserver.com kubelet[20716]:       --stderrthreshold severity                            logs at or above this
threshold go to stderr (default 2)
Oct 29 12:35:36 dworker1.mylabserver.com kubelet[20716]:   -v, --v Level                                             log level for V logs
Oct 29 12:35:36 dworker1.mylabserver.com kubelet[20716]:       --version version[=true]                              Print version informat
ion and quit
Oct 29 12:35:36 dworker1.mylabserver.com kubelet[20716]:       --vmodule moduleSpec                                  comma-separated list o
f pattern=N settings for file-filtered logging
Oct 29 12:35:36 dworker1.mylabserver.com systemd[1]: kubelet.service: Failed with result 'exit-code'.
Oct 29 12:35:36 dworker1.mylabserver.com kubelet[20716]:       --volume-plugin-dir string                            The full path of the d
irectory in which to search for additional third party volume plugins (default "/usr/libexec/kubernetes/kubelet-plugins/volume/exec/")
Oct 29 12:35:36 dworker1.mylabserver.com kubelet[20716]: F1029 12:35:36.741989   20716 server.go:145] unknown flag: --ignore-preflight-erro
rs

```

**Comment by aweknee** [comment]
I was able to make KTHW work with Docker CE instead of containerd.

This is what I did on one of the worker nodes:
- Stopped and disabled the running containerd, kubelet, and kube-proxy services.
- Verified that that node got into NotReady state.
- Installed Docker CE.
- Altered the docker daemon service to use this command line and restarted it to take effect:
```
  ExecStart=/usr/bin/dockerd -H fd:// --containerd=/run/containerd/containerd.sock \
   --iptables=false --ip-masq=false
``` 
- Changed the kubelet service's command line to:
```
ExecStart=/usr/local/bin/kubelet \
  --config /var/lib/kubelet/kubelet-config.yaml \
  --container-runtime docker \
  --image-pull-progress-deadline 2m \
  --kubeconfig /var/lib/kubelet/kubeconfig \
  --network-plugin cni \
  --register-node="true" \
  --v 2
```
- Enabled and started the kubelet and kube-proxy services.
- Verified that the node is back to Ready state and is able to run pods scheduled to it.

**Comment by kelseyhightower** [comment]
I'm glad this works but it's a personal preference to keep containerd as the runtime.

**Comment by andrei-matei** [comment]
I have been having issues with CoreDNS where I got really slow times resolving service names. It took up to 30 seconds. After a lot of debugging, I took out containerd and used docker as a runtime and dns works perfect. I really cannot understand why this happened. 
Now I want to try with CRI-O. I really cannot understand why the container runtime has anything to do with dns resolution but I suspect some kind of loop in domains from resolv.conf.
